Sources/Bodega/SQLiteStorageEngine+SQLiteAccess.swift

```diff
@@ -0,0 +1,303 @@
+            \(Column.updatedAt.rawValue) = EXCLUDED.\(Column.updatedAt.rawValue)
+        """
+        var insertStatement: OpaquePointer?
+        try prepareDatabase(database, forCommand: command, statementPointer: &insertStatement)
+        let currentDateString = Self.dateFormatter.string(from: Date())
+        for entry in dataAndKeys {
+            let bytes = [UInt8](entry.data)
+            sqlite3_bind_text(insertStatement, 1, strdup(entry.key.rawValue), -1, SQLITE_TRANSIENT)
+            sqlite3_bind_blob(insertStatement, 2, bytes, Int32(bytes.count), SQLITE_TRANSIENT)
+            sqlite3_bind_text(insertStatement, 3, strdup(currentDateString), -1, SQLITE_TRANSIENT)
+            sqlite3_bind_text(insertStatement, 4, strdup(currentDateString), -1, SQLITE_TRANSIENT)
+            let result = sqlite3_step(insertStatement)
+            guard result == SQLITE_DONE else {
+                throw SQLiteError.insertFailed(result)
+            }
+            sqlite3_reset(insertStatement)
+        }
+        sqlite3_finalize(insertStatement)
+    }
+
+    // MARK: - DELETE
+
+    static func deleteKeys(_ keys: [CacheKey], fromTable tableName: String, inDatabase database: OpaquePointer?) throws {
+        let command = """
+        DELETE FROM
+            \(tableName)
+        WHERE
+            \(Column.key.rawValue) IN (\(Array(repeating: "?", count: keys.count).joined(separator: ", ")))
+        LIMIT
+            \(keys.count)
+        """
+        var removeStatement: OpaquePointer?
+        try prepareDatabase(database, forCommand: command, statementPointer: &removeStatement)
+        for (index, key) in keys.enumerated() {
+            sqlite3_bind_text(removeStatement, Int32(index + 1), strdup(key.rawValue), -1, SQLITE_TRANSIENT)
+        }
+        let result = sqlite3_step(removeStatement)
+        guard result == SQLITE_DONE else {
+            throw SQLiteError.deleteFailed(result)
+        }
+        sqlite3_finalize(removeStatement)
+    }
+
+    static func emptyTableNamed(_ tableName: String, inDatabase database: OpaquePointer?) throws {
+        let command = """
+        DELETE FROM
+            \(tableName)
+        """
+        var deleteStatement: OpaquePointer?
+        try prepareDatabase(database, forCommand: command, statementPointer: &deleteStatement)
+        let result = sqlite3_step(deleteStatement)
+        guard result == SQLITE_DONE else {
+            throw SQLiteError.deleteFailed(result)
+        }
+        sqlite3_finalize(deleteStatement)
+    }
+
+    // MARK: - SELECT
+
+    static func selectKeys(_ keys: [CacheKey], inTable tableName: String, inDatabase database: OpaquePointer?) -> [(key: CacheKey, data: Data)] {
+        let command = """
+        SELECT
+            \(Column.key.rawValue), \(Column.value.rawValue)
+        FROM
+            \(tableName)
+        WHERE
+            \(Column.key.rawValue) IN (\(Array(repeating: "?", count: keys.count).joined(separator: ", ")))
```

Comment:
This is OK, but not exactly cheap. This creates a ton of malloced objects just to concat a static String. Maybe rather make that procedural.

Also again, is the key guaranteed to be SQL safe or does it need escaping? Aka proper binding.